### PR TITLE
cairo: simplifications for Linuxbrew

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

**Comment by maxim-belkin** [comment]
One of the failed tests:
```
The formula built, but is not symlinked into /home/linuxbrew/.linuxbrew
Could not symlink share/man/man3/uuid.3
Target /home/linuxbrew/.linuxbrew/share/man/man3/uuid.3
is a symlink belonging to util-linux. You can unlink it:
  brew unlink util-linux

To force the link and overwrite all conflicting files:
  brew link --overwrite ossp-uuid

To list all files that would be deleted:
  brew link --overwrite --dry-run ossp-uuid
```

**Comment by maxim-belkin** [comment]
One interesting thing I discovered: if `valgrind` is installed, `cairo`'s build system detects it and tries to use it. Installation, however, fails because Homebrew superenv "hides" `valgrind`'s include directories because `valgrind` is not a dependency (at the moment) of `cairo`. Here I'm adding `valgrind` as a recommended dependency. We shall discuss if it should be optional.

**Comment by maxim-belkin** [comment]
@sjackman, shall we remove uuid.3 man page from util-linux? (see https://github.com/Linuxbrew/homebrew-core/pull/9441#issuecomment-422380872)

**Comment by maxim-belkin** [comment]
Forgot to explain: `cairo` is an optional dependency of `libdrm`... That's why I'm here

**Comment by maxim-belkin** [comment]
The other failure is:
```
==> brew linkage --test gtk+
Missing libraries:
  libXinerama.so.1
==> FAILED
```

**Comment by sjackman** [comment]
`ossp-uuid` and `util-linux` conflict. `ossp-uuid` should probably be `keg_only` on Linux.

Whichever formula is using `ossp-uuid` on Linux should be changed to instead use `util-linux`.

File: Formula/cairo.rb [hunk 1]
```diff
@@ -34,33 +33,25 @@ class Cairo < Formula
...
     depends_on "zlib"
-    depends_on "linuxbrew/xorg/xorg" => :recommended
-    if build.with?("xorg")
-      depends_on "linuxbrew/xorg/kbproto"
-      depends_on "linuxbrew/xorg/libxcb"
-      depends_on "linuxbrew/xorg/renderproto"
-      depends_on "linuxbrew/xorg/xcb-proto"
-      depends_on "linuxbrew/xorg/xextproto"
-    end
+    depends_on "valgrind" => :recommended
```
Comment (sjackman) [inline Formula/cairo.rb]: Please remove this option.

File: Formula/cairo.rb [hunk 2]
```diff
@@ -34,33 +33,25 @@ class Cairo < Formula
...
       --disable-dependency-tracking
       --prefix=#{prefix}
       --enable-gobject=yes
       --enable-svg=yes
       --enable-tee=yes
+      --enable-quartz-image=#{OS.mac? ? "yes" : "no"}
+      --enable-xcb=#{notmac}
+      --enable-xlib=#{notmac}
+      --enable-xlib-xrender=#{notmac}
+      --enable-valgrind=#{build.with?("valgrind") ? "yes" : "no"}
```
Comment (sjackman) [inline Formula/cairo.rb]: Please change this line to `--enable-valgrind=no`

File: Formula/gtk+.rb [hunk 3]
```diff
@@ -33,6 +33,7 @@ class Gtkx < Formula
   depends_on "pango"
   depends_on "hicolor-icon-theme"
   depends_on "cairo" unless OS.mac?
+  depends_on "linuxbrew/xorg/libxinerama" unless OS.mac?
```
Comment (sjackman) [inline Formula/gtk+.rb]:
```ruby
unless OS.mac?
  depends_on "cairo"
  depends_on "linuxbrew/xorg/libxinerama"
end
```
Is `libxinerama` necessary?

File: Formula/cairo.rb [hunk 4]
```diff
@@ -34,33 +33,25 @@ class Cairo < Formula
...
     args = %W[
       --disable-dependency-tracking
       --prefix=#{prefix}
       --enable-gobject=yes
       --enable-svg=yes
       --enable-tee=yes
+      --enable-quartz-image=#{OS.mac? ? "yes" : "no"}
+      --enable-xcb=#{notmac}
+      --enable-xlib=#{notmac}
+      --enable-xlib-xrender=#{notmac}
```
Comment (sjackman) [inline Formula/cairo.rb]: I know it seems a bit repetitive, but I prefer not to factor out `notmac`.
```
--enable-xcb=#{OS.mac? ? "no" : "yes"}
…
```

File: Formula/cairo.rb [hunk 5]
```diff
@@ -4,14 +4,13 @@ class Cairo < Formula
   url "https://cairographics.org/releases/cairo-1.14.12.tar.xz"
   mirror "https://www.mirrorservice.org/sites/ftp.netbsd.org/pub/pkgsrc/distfiles/cairo-1.14.12.tar.xz"
   sha256 "8c90f00c500b2299c0a323dd9beead2a00353752b2092ead558139bd67f7bf16"
-  revision 3 unless OS.mac?
+  revision 4 unless OS.mac?
```
Comment (sjackman) [inline Formula/cairo.rb]: There should be no need to bump the revision nor update the bottle.

File: Formula/cairo.rb [hunk 6]
```diff
@@ -4,14 +4,13 @@ class Cairo < Formula
...
 
   bottle do
     sha256 "5a6cc135f8a373376dac7d8e2750d10c955fd83977f5549976ad590958971f93" => :mojave
     sha256 "5bdc28de8e5a615ab664d43f7f322ed02d58071171415bb6e2750f486b9465e2" => :high_sierra
     sha256 "102847d74a0a11bb6143d93b9f32e1736e88036fb4c685d554a8bcd376bbd929" => :sierra
     sha256 "bec85433a35605164bdbf5f8913e29eb6d9ceb5acc5569dd9d864706ae6c8d49" => :el_capitan
-    sha256 "f280c34cad69a036fb1f209249e054e78da643481afd8a3962bed5ad91ca9ccb" => :x86_64_linux
```
Comment (sjackman) [inline Formula/cairo.rb]: Ditto

File: Formula/gtk+.rb (hunk 3 above)
Comment (maxim-belkin) [reply Formula/gtk+.rb]: https://github.com/Linuxbrew/homebrew-core/pull/9441#issuecomment-422481556

File: Formula/cairo.rb (hunk 2 above)
Comment (maxim-belkin) [reply Formula/cairo.rb]: ok

File: Formula/cairo.rb (hunk 1 above)
Comment (maxim-belkin) [reply Formula/cairo.rb]: ok

File: Formula/cairo.rb (hunk 6 above)
Comment (maxim-belkin) [reply Formula/cairo.rb]: k

File: Formula/cairo.rb (hunk 5 above)
Comment (maxim-belkin) [reply Formula/cairo.rb]: k

File: Formula/cairo.rb (hunk 4 above)
Comment (maxim-belkin) [reply Formula/cairo.rb]: k

File: Formula/gtk+.rb (hunk 3 above)
Comment (maxim-belkin) [reply Formula/gtk+.rb]:
```
$ brew linkage gtk+
...
  /home/linuxbrew/.linuxbrew/lib/libXinerama.so.1 (linuxbrew/xorg/libxinerama)
```

**Comment by sjackman** [comment]
```
$ brew linkage --reverse gtk+ | grep -A2 libXinerama
/home/linuxbrew/.linuxbrew/lib/libXinerama.so.1
  lib/libgdk-x11-2.0.so.0.2400.32
```
So looks like indeed a necessary dependency.

**Comment by sjackman** [comment]
Looks good to me for a squash merge once CI is 🍏
Thanks, Maxim!

**Comment by maxim-belkin** [comment]
I don't think it will be green until `uuid` issue is fixed

**Comment by sjackman** [comment]
Failures are…
```
==> brew linkage --test io
==> FAILED
==> Installing neko dependency: httpd
==> FAILED
==> brew linkage --test wxmac
==> FAILED
```
https://circleci.com/gh/Linuxbrew/homebrew-core/20412

which are unrelated to this PR. I'm okay with squash-merging this PR as is. It'd be great if you wanted to address these three failures, each in a separate PR.

**Comment by xu-cheng** [comment]
After this PR, `xorg` becomes hard dependency for `cairo`, is it possible to restore the option of `--without-xorg`(I can send a PR)?

**Comment by maxim-belkin** [comment]
Hey, @xu-cheng. My apologies if it broke something for you. 

> is it possible to restore the option of --without-xorg

I don't see why not. I'd like to point out that after this PR `cairo` depends only on a subset of packages from `linuxbrew/xorg` (16 to build from source, 11 to pour a bottle), not the entire `xorg`

**Comment by xu-cheng** [comment]
> I'd like to point out that after this PR cairo depends only on a subset of packages from linuxbrew/xorg (16 to build from source, 11 to pour a bottle), not the entire xorg

Thanks for the info. I think that would work for me.

FYI, the reason that I want to avoid `xorg` is because my system is headless. Therefore, it is unnecessary to install `xorg`. Further, installing some dependencies of `xorg` is cumbersome and error-prone to me.

**Comment by sjackman** [comment]
So long as nothing is broken, I suggest accepting the `xorg` dependency. If you are getting build failures with some of the components of `xorg`, let's address that issue.

**Pull request closed without merging**